Mathlib/Combinatorics/SimpleGraph/Extremal/Basic.lean

```diff
@@ -0,0 +1,167 @@
+/-
+Copyright (c) 2025 Mitchell Horner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mitchell Horner
+-/
+import Mathlib.Algebra.Order.Floor
+import Mathlib.Combinatorics.SimpleGraph.Operations
+import Mathlib.Combinatorics.SimpleGraph.Copy
+
+/-!
+# Extremal graph theory
+
+This file introduces basic definitions for extremal graph theory, including extremal numbers.
+
+## Main definitions
+
+* `SimpleGraph.extremalNumber` is the maximum number of edges in a `H`-free simple graph on `n`
+  vertices.
+
+  If `H` is contained in all simple graphs on `n` vertices, then this is `0`.
+
+* `SimpleGraph.IsExtremal` is the predicate that `G` satisfies `p` and any `H` satisfying `p` has
+  at most as many edges as `G`.
+-/
+
+
+open Finset Fintype
+
+namespace SimpleGraph
+
+section ExtremalNumber
+
+open Classical in
+/-- The extremal number of a natural number `n` and a simple graph `H` is the the maximum number of
+edges in a `H`-free simple graph on `n` vertices.
+
+If `H` is contained in all simple graphs on `n` vertices, then this is `0`. -/
+noncomputable def extremalNumber (n : ℕ) {W : Type*} (H : SimpleGraph W) : ℕ :=
+  sup { G : SimpleGraph (Fin n) | H.Free G } (#·.edgeFinset)
+
+variable {n : ℕ} {V W : Type*} {G : SimpleGraph V} {H : SimpleGraph W}
+
+open Classical in
+theorem extremalNumber_eq_sup_overFin {n : ℕ} [Fintype V] (hc : card V = n) :
+    extremalNumber n H = sup { G : SimpleGraph V | H.Free G } (#·.edgeFinset) := by
+  let e := Fintype.equivFinOfCardEq hc
+  rw [extremalNumber, le_antisymm_iff]
+  and_intros
+  on_goal 1 =>
+    replace e := e.symm
+  all_goals
+  rw [Finset.sup_le_iff]
+  intro G h
+  let G' := G.map e.toEmbedding
+  replace h' : G' ∈ univ.filter (H.Free ·) := by
+    rw [mem_filter, ← free_congr Iso.refl (Iso.map e G)]
+    simpa using h
+  rw [Iso.card_edgeFinset_eq (Iso.map e G)]
+  convert @le_sup _ _ _ _ { G | H.Free G } (#·.edgeFinset) G' h'
+
+variable [Fintype V] [DecidableRel G.Adj]
+
+/-- If `G` is `H`-free, then `G` has at most `extremalNumber (card V) H` edges. -/
+theorem le_extremalNumber (h : H.Free G) : #G.edgeFinset ≤ extremalNumber (card V) H := by
```

Comment:
```suggestion
theorem card_edgeFinset_le_extremalNumber (h : H.Free G) : #G.edgeFinset ≤ extremalNumber (card V) H := by
```
maybe? Even better might be to have an `abbrev` for `#G.edgeFinset`. It's a very common idiom of course